File: Bridge/c++/Bridge.cpp
#include <iostream>

class DataReader
{
    public:
        virtual void read() = 0;
};

class DataBaseReader : public DataReader
{
    public:
        void read() override;
};

class FileReader : public DataReader
{
    public:
        void read() override;
};

class Sender
{
    protected:
        DataReader* reader;

    public:
        Sender(DataReader* dataReader) : reader(dataReader) {}
        void SetDataReader(DataReader* dataReader);
        virtual void send() = 0;
};

class EmailSender : public Sender
{
    public:
        EmailSender(DataReader* dataReader) : Sender(dataReader) {}
        void send() override;
};

class TelegramBotSender : public Sender
{
    public:
        TelegramBotSender(DataReader* dataReader) : Sender(dataReader) {}
        void send() override;
};

int main()
{
    Sender* sender = new EmailSender(new DataBaseReader());
    sender->send();

    sender->SetDataReader(new FileReader());
    sender->send();

    sender = new TelegramBotSender(new DataBaseReader());
    sender->send();

    return 0;
}

void DataBaseReader::read()
{
    std::cout << "data from the database ";
}

void FileReader::read()
{
    std::cout << "data from the file ";
}

void Sender::SetDataReader(DataReader *dataReader)
{
    reader = dataReader;
}

void EmailSender::send()
{
    reader->read();
    std::cout << "sent by email\n";
}

void TelegramBotSender::send()
{
    reader->read();
    std::cout << "sent by telegram bot\n";
}

File: Bridge/python/Bridge.py
import abc


class IDataReader(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def read(self):
        pass


class DataBaseReader(IDataReader):

    def read(self):
        print("data from the database ", end='')


class FileReader(IDataReader):

    def read(self):
        print("data from the file ", end='')


class Sender(metaclass=abc.ABCMeta):

    def __init__(self, data_reader: IDataReader):
        self.reader: IDataReader = data_reader

    def set_data_reader(self, data_reader: IDataReader):
        self.reader: IDataReader = data_reader

    @abc.abstractmethod
    def send(self):
        pass


class EmailSender(Sender):

    def __init__(self, data_reader: IDataReader):
        super().__init__(data_reader)

    def send(self):
        self.reader.read()
        print("sent by email")


class TelegramBotSender(Sender):

    def __init__(self, data_reader: IDataReader):
        super().__init__(data_reader)

    def send(self):
        self.reader.read()
        print("sent by telegram bot")


if __name__ == '__main__':

    sender: Sender = EmailSender(DataBaseReader())
    sender.send()

    sender.set_data_reader(FileReader())
    sender.send()

    sender = TelegramBotSender(DataBaseReader())
    sender.send()